src/messages/accumulating.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{DstLocation, Message, SrcAuthority, Variant};
+use crate::{
+    chain::{SectionKeyShare, SectionProofSlice},
+    crypto::{self, Digest256},
+    error::Result,
+    xor_space::{Prefix, XorName},
+};
+use bincode::serialize;
+use log::LogLevel;
+use std::{collections::BTreeSet, mem};
+
+/// Section-source message that is in the process of signature accumulation.
+/// When enough signatures are collected, it can be converted into full `Message` by calling
+/// `combine_signatures`.
+#[allow(missing_docs)]
+#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+pub struct AccumulatingMessage {
+    pub src: Prefix<XorName>,
+    pub dst: DstLocation,
+    pub variant: Variant,
+    pub proof: SectionProofSlice,
+    pub public_key_set: bls::PublicKeySet,
+    pub signature_shares: BTreeSet<(usize, bls::SignatureShare)>,
+}
+
+impl AccumulatingMessage {
+    /// Create new `AccumulatingMessage`
+    pub fn new(
+        src: Prefix<XorName>,
+        dst: DstLocation,
+        variant: Variant,
+        section_share: &SectionKeyShare,
+        public_key_set: bls::PublicKeySet,
+        proof: SectionProofSlice,
+    ) -> Result<Self> {
+        let bytes = serialize(&(&src, &dst, &variant))?;
+        let mut signature_shares = BTreeSet::new();
+        let sig_share = section_share.key.sign(&bytes);
+        let _ = signature_shares.insert((section_share.index, sig_share));
+
+        Ok(Self {
+            src,
+            dst,
+            variant,
+            proof,
+            public_key_set,
+            signature_shares,
+        })
+    }
+
+    /// Add the signature shares of `other` into this message.
+    /// Note: currently no validation is performed that the messages have the same content.
```

Comment:
Perhaps a big FIXME here to be addressed soon and we can look t verifying the content and sig at the same time and do away with a bad_sigs container.

Reply:
I'd say this is a low priority because in practice the signature accumulator calls this function only on messages that have identical hashes (and thus content). Might be still worth at least raising a github issue for it.